### ISSUE TITLE
async tracing_subscriber::layer::Layer on_event

### ISSUE THREAD

**Issue by LukeMauldin** [tokio-rs/tracing#738]
### Crates
tracing-subscriber

### Motivation

I have my own struct implementing `tracing_subscriber::layer::Layer - on_event` so if the event is an `error` that I can send an HTTP request to log the error.  This was all working fine before I upgraded to the async version of `reqwest`.  After I did that, I have some code that looks like this to try to log the message:
```
 if let Ok(handle) = tokio::runtime::Handle::try_current() {
            //No need to wait on the spawn, the tokio runtime will handle pooling the future to make sure it is completed
            handle.spawn(async move {
                let res = webutil::send_request_no_resp(&uri, &record, None).await;

                match res {
                    Ok(()) => (),
                    Err(e) => {
                        println!(
                            "error: unable to log message: '{}' due to error: {}",
                            record.Message, e
                        );
                    }
                };
            });
        } else {
            //Handle logging without being in the context of an existing tokio runtime
            let res = tokio::runtime::Builder::new()
                .basic_scheduler()
                .enable_all()
                .build()
                .unwrap()
                .block_on(webutil::send_request_no_resp(&uri, &record, None));
            match res {
                Ok(()) => (),
                Err(e) => {
                    println!(
                        "error: unable to log message: '{}' due to error: {}",
                        record.Message, e
                    );
                }
            };
        }
```

The code above is not ideal and although it passes my unit tests, I am not sure it is correct for every situation.  It would be more convenient if somehow the crate would expose an `async` method for processing `on_event`.  I am a relatively new to rust and I am not sure how to recommend a solution.

**Comment by davidbarsky** [comment]
## Our Current Thinking

This question has a bit overlap with #668. I recommend reading that issue first and [Eliza's comment](https://github.com/tokio-rs/tracing/issues/668#issuecomment-609945964) in particular. That will explain why `tracing_subscriber::layer::Layer` is designed the way that it is and why I _personally_ don't think it'll change to support asynchronous functions.

## A Solution

I suggest looking at `tracing-appender`, or if that doesn't work, using some asynchronous channel (like [`tokio::sync`](https://docs.rs/tokio/0.2.21/tokio/sync/mpsc/index.html) whose `try_send` methods are _not_ asynchronous and would be considered non-blocking. A background task spawned by `tokio::spawn` can be can read off this channel and make the `webutil::send_request_no_resp` calls that you have in your code.

**Comment by hawkw** [comment]
Unfortunately, it isn't really possible to have `async fn` methods in the `Layer` trait. This is for a few reasons. 

First of all, [functions in traits cannot currently be declared `async`](https://doc.rust-lang.org/stable/error-index.html#E0706), so if we wanted to do this, these methods would have to return futures, which introduces complexity to the interface.

However, even if it does become possible to provide `async fn` versions of these methods, I don't think it would make sense to here. Async functions are only executed when `.await`ed in an asynchronous context. This means that the `tracing` macros would not be callable in synchronous functions, as they would either expand to include an `await` or have to be awaited, or would spawn a task using a runtime-specific API. This would, of course, impact the use of `tracing` in synchronous code. Even if we didn't care about supporting non-async projects, though, using `.await` in the `tracing` macros would impact asynchronous programs as well: async programs almost always also contain some synchronous functions that are called to perform some logic that does not involve IO. For example, if you wrote code like this
```rust
fn do_some_business_logic(thing: Thing) -> OtherThing { 
    tracing::info!(?thing, "doing some business logic to a Thing").await;
    // ...
}

async fn turn_thing_into_other_thing() {
    let thing = get_thing().await;
    let other_thing = do_some_business_logic(thing);
    // ...
}
```
this wouldn't work, since even though `do_some_business_logic` is called in an asynchronous context, it is not _itself_ async, and we cannot `.await` the `tracing::info!` macro there. This would also be the case if the macros contained the `.await`. The other approach, of spawning a task, _could_ work in this case, but APIs for spawning are currently runtime-specific — if we did something like the example code you posted, `tracing` would only work in projects that use Tokio, which would not be ideal.

So, what can we do instead? 

The approach you've posted above does indeed work, since the `Layer` implementation itself is not part of the core `tracing` library, and _can_ make assumptions like "I will be running on a Tokio runtime". However, spawning a new task for every event may not be the most efficient solution, since events may occur very frequently; it also doesn't necessarily guarantee that the events are reported in order, since tasks are not necessarily executed in the order that they are spawned. Since you're only making HTTP requests on errors, rather than for all events that occur, your approach might be fine if the errors do not occur very frequently.

It might be preferable to do something more like what @davidbarsky suggested in https://github.com/tokio-rs/tracing/issues/738#issuecomment-635487656, and use a channel to send events to a dedicated background task or thread. The `tracing-appender` crate's `non_blocking` writer uses a dedicated background _thread_, rather than a spawned task. This is partially because it may not necessarily be used in async programs. Additionally, since it records _all_ events that occur, it needs to cope with a pretty high volume of events; the background thread will probably be pretty busy, and this avoids tying up an async worker thread for a long period of time. However, we could also provide an async version in the future.

Something else we might want to experiment with is providing a _separate trait_ for an asynchronous layer, possibly using the [`async-trait`](https://docs.rs/async-trait/0.1.31/async_trait/index.html) crate. We could provide a `Layer` implementation that takes a trait implementing an "async layer", and sends the futures it produces to a dedicated background task to execute. This interface could allow writing custom layers using `async fn`, and would let layer authors avoid the boilerplate of managing a dedicated background task. I'm not sure how well this idea would work out in practice, but it's a thought, and might be worth investigating.

Does that all make sense? Let me know if you have any further questions!
